APE_Black_Formatting.rst

```diff
@@ -0,0 +1,286 @@
+allowed if the code author and PR reviewer(s) agree that the code line or block
+should be excluded from Black formatting.**
+
+
+Branches and pull requests
+--------------------------
+
+- `Add configuration for black autoformatter <https://github.com/astropy/astropy/pull/13253>`_
+- `Apply black to modeling <https://github.com/astropy/astropy/pull/13254>`_
+
+Implementation
+--------------
+
+All Python code in ``astropy`` is formatted with Black_, using its default
+settings, that is, 88 characters per line and double quotes. Note that, by
+design, most of Black_ is inflexible, and there is not a way to pick-and-choose
+some elements and not others of Black's format.
+
+Implementing this change requires:
+
+1. Updating the `coding style`_:
+
+   * Adding documentation about Black, with a particular emphasis on the rationale expressed here and the exception policy described above.
+   * Adding documentation about using pre-commit, ala `pandas pre-commit`_.
+   * Document the black exception guidelines (as described in this APE)
+   * Removing other references to code formatting in the documentation.
+
+2. Updating flake8 configuration to be compatible with Black.
+
+   This is straightforward and well documented. It may even be removed if it
+   does not provide any added checks that ``black –check`` does not.
+
+3. Checking Black in CI.
+
+   * Add a CI check that runs ``black --check`` on the whole code base, failing
+     if it meets a violation
+   * Allow for the Black CI bot to fix formatting errors via documented single
+     line comments. This is an opt-in step which could be used by maintainers to
+     fix errors prior to merging. Having such an option available could lower
+     the barrier for new contributors.
+
+4. Reformatting Astropy's code.  This will be done on a subpackage-by-subpackage
+   basis over a transition period. The process for each subpackage to be
+   transitioned is as follows:
+
+   * The sub package maintainer(s) create a branch (e.g.
+     ``blackify-coordinates``) to do the re-formatting.
+   * Edit ``pyproject.toml`` to remove the black exclusion for the sub-package.
+   * Run ``black --skip-string-normalization astropy/<sub-package>`` on the
+     sub-package directory.  Commit the changes black makes and make a draft
+     pull request.
+   * The maintainer(s) should review all changes and highlight for discussion
+     any  changes that are judged to be valid exceptions (see above for what
+     "valid exception" means). These can be reverted by copying-and-pasting
+     from the "changes" interfaces to the local copy and applying the
+     appropriate Black directives to prevent auto-formatting. Commit and push
+     these changes for review.
+   * Once the PR submitter and reviewer are both satisfied with the changes
+     then squash the commit(s) down to a single commit and force push.
+   * Run ``black astropy/<sub-package>`` to apply just the string normalization
```

Comment:
Note that string normalization converts the quotes around all strings to be `"` unless the string in question contains a `"` character.